### ISSUE TITLE
Implied "else" narrowing doesn't work with local variable

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1844]
**Describe the bug**
If the variable being narrowed is not passed as an argument but defined locally, implied "else" narrowing does not work.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That no error would be produced.

**Screenshots or Code**

```python
from typing import Literal


def func2(x: Literal[1, 2]):
    if x == 1 or x == 2:
        y = True

    print(y) # No error


def func3():
    x = 1 if 1 else 2
    reveal_type(x)
    if x == 1 or x == 2:
        y = True

    print(y) # "y" is possibly unbound
```

**VS Code extension or command-line**
1.1.138

**Comment by erictraut** [comment]
This behavior isn't specific to local variables. It's specific to expressions that have no declared types. In your first example (`func2`), the expression `x` has a declared type. In your second example (`func3`), the type of `x` must be inferred.

The implied "else" narrowing is a very expensive operation, so we needed to impose some limitations. One such limitation is that the reference expression (in this case `x`) must have a declared type. You'll notice that it works if you add the type annotation `x: Literal[1, 2]`. 

I don't see a good way to eliminate this limitation without significantly slowing down analysis times, and that's not a tradeoff we're willing to make.

**Comment by layday** [comment]
Okay, thank you for explaining.  This was happening to me with a class attribute which was annotated, which is why I did not think that the absence of a declared type was the culprit:

```py
from typing import Literal

class Foo:
    bar: Literal[1, 2]

def baz():
    foo = Foo()
    if foo.bar == 1 or foo.bar == 2:
        y = True

    print(y) # "y" is possibly unbound
```

Assigning the attribute to a local variable and re-annotating it made the error go away.
